### PR TITLE
Support async server errors

### PR DESCRIPTION
**Pull request by brodin**
This will add the possibility to send async errors back as a response from the server handler while staying backwards compatible with the current callback style. To get the test working I fixed a current bug where the tests in `request-response-samples-test.js` exposed its before/after/beforeEach hooks in a way which applied them to every other test in the suite.

### CODE REVIEW

**Comment by brodin** [comment]
@herom any idea why these unrelated tests fails? Works locally with the different versions of node.

**Comment by simonjosefsson** [comment]
The first two errors are mocha timeouts, probably caused by peak demand on the build system
(FYI, one can increase mocha's defaults two second timeout with --timeout <millisecond>)

The third one could be 0.12 specific, or a fluke... hard to say

**Comment by brodin** [comment]
@herom changed this to one commit rather than two now.

File: test/request-response-samples-test.js [hunk 1]
```diff
@@ -36,31 +36,6 @@ var requestContext = {
   }
 };
 
-var origRandom = Math.random;
-module.exports = {
```
Comment (herom) [inline test/request-response-samples-test.js]: Why was this removed?

Comment (brodin) [inline test/request-response-samples-test.js]: Tried to explain the purpose in my commit message:
> Changed how tests are generated to avoid global before/after hooks. By using the format with exporting before, beforeEach and after hooks aswell as the test suite one actually register the hooks globally - applying them to every test run by npm test. This wasn't really a problem until you try to do setTimeouts, which will never resolve since timekeeper.freeze is used (https://github.com/vesln/timekeeper#featuresproblems)"

Before it was attached to the global before hook, which meant that every test in the entire suite would get a fake random - if they wanted it or not. Now I limited the fake random to just be for the "request-response-sample" tests. The biggest problem was with timekeeper, blowing up any attempt to use setTimeouts in a test.

Comment (herom) [inline test/request-response-samples-test.js]: Thanks a lot for the explanation - just wanted to be sure :+1:

**Comment by gangstead** [comment]
As mentioned in #613 after this fix a http status code of 200 is still returned even when there is a soap fault.
